File: leetcode/cpp/stack/lc_150_evaluate_reverse_polish_notation.cpp
#include <stack>
#include <vector>
#include <string>

/**
 * https://leetcode.com/problems/evaluate-reverse-polish-notation/description/
 *
 * Time complexity: O(n) where n = length of tokens
 * Space complexity: O(n) where n = length of tokens
 */
class Solution {
public:
    int evalRPN(std::vector<std::string>& tokens) {
        // init our stack to store our calculations
        std::stack<int> calculations;

        for (const std::string& token: tokens) {
            // if we get an operator token then
            // we can get the first and second nums
            // from our calculations stack
            // perform the operations and put the result
            // back onto the stack
            if (token == "+") {
                int secondNum = calculations.top();
                calculations.pop();
                int firstNum = calculations.top();
                calculations.pop();
                calculations.push(firstNum + secondNum);
            } else if (token == "-") {
                int secondNum = calculations.top();
                calculations.pop();
                int firstNum = calculations.top();
                calculations.pop();
                calculations.push(firstNum - secondNum);
            } else if (token == "*") {
                int secondNum = calculations.top();
                calculations.pop();
                int firstNum = calculations.top();
                calculations.pop();
                calculations.push(firstNum * secondNum);
            } else if (token == "/") {
                int secondNum = calculations.top();
                calculations.pop();
                int firstNum = calculations.top();
                calculations.pop();
                calculations.push(firstNum / secondNum);
            // else if we get a number we can simply add it
            // to the calculations stack
            } else {
                calculations.push(std::stoi(token));
            }
        }

        // finally, we return the last number in our calculations stack
        return calculations.top();
    }
};

File: leetcode/stacks/lc_150_evaluate_reverse_polish_notation.py
from typing import List


'''
https://leetcode.com/problems/evaluate-reverse-polish-notation/description/
'''
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        # create a set of the operands 
        # that can be in the tokens array
        operands = {'+', '-', '*', '/'}
        
        # init our stack
        stack = []

        # loop through all tokens
        for token in tokens:
            # check if the token is an operand
            if (token in operands):
                # get the second number from the stack
                second_num = stack.pop()
                # get the first number from the stack
                first_num = stack.pop()

                # create the math expression we want to evaluate
                math_expression = f"{first_num} {token} {second_num}"
                # calculate the math expression using eval()
                calculation = eval(math_expression)
                
                # add the calculation back to the stack
                # use int() to truncate towards 0
                stack.append(int(calculation))

            # else the token must be a number
            else:
                # add the number to the stack
                # use int() to handle the edge case
                # where we are given a single number
                # and don't want to return a string number
                stack.append(int(token))
        
        # return the final number in the stack
        # after all calculations are complete
        return stack[-1]
